crates/sui-rosetta/src/types/internal_operation/pay_sui.rs

```diff
@@ -0,0 +1,147 @@
+    pub amounts: Vec<u64>,
+}
+
+#[async_trait]
+impl TryConstructTransaction for PaySui {
+    async fn try_fetch_needed_objects(
+        self,
+        client: &SuiClient,
+        gas_price: Option<u64>,
+        budget: Option<u64>,
+    ) -> Result<TransactionObjectData, Error> {
+        let Self {
+            sender,
+            recipients,
+            amounts,
+        } = self;
+
+        let total_amount = amounts.iter().sum::<u64>();
+        if let Some(budget) = budget {
+            // We have a constant budget, so no need to dry-run
+            let all_coins = client
+                .coin_read_api()
+                .select_coins(sender, None, (total_amount + budget) as u128, vec![])
+                .await?;
+
+            let total_sui_balance = all_coins.iter().map(|c| c.balance).sum::<u64>() as i128;
+
+            let mut iter = all_coins.into_iter().map(|c| c.object_ref());
+            let gas_coins: Vec<_> = iter.by_ref().take(MAX_GAS_COINS).collect();
+            let extra_gas_coins: Vec<_> = iter.collect();
+
+            return Ok(TransactionObjectData {
+                gas_coins,
+                extra_gas_coins,
+                objects: vec![],
+                total_sui_balance,
+                budget,
+            });
+        };
+
+        let gas_price = match gas_price {
+            Some(p) => p,
+            None => client.governance_api().get_reference_gas_price().await? + 100, // make sure it works over epoch changes
+        };
+
+        let mut coins_stream = Box::pin(client.coin_read_api().get_coins_stream(sender, None));
+
+        let mut all_coins = vec![];
+        let mut gas_coins: Vec<_>;
+        let mut extra_gas_coins: Vec<_>;
+        let total_amount = amounts.iter().sum::<u64>();
+        let mut gathered = 0;
+        let mut budget = START_GAS_UNITS * gas_price;
+        // We need to dry-run in a loop, because depending on the amount of coins used the tx might
+        // differ slightly: (merge / no merge / number of merge-coins)
+        loop {
+            while let Some(coin) = coins_stream.next().await {
+                gathered += coin.balance;
+                all_coins.push(coin);
+                if gathered >= total_amount + budget {
+                    break;
+                }
+            }
+            if gathered < total_amount + budget {
+                return Err(Error::InvalidInput(
+                    format!("Address {sender} does not have enough Sui balance to transfer {total_amount} with needed budget: {budget}. Sui balance: {gathered}.")
+                ));
+            }
+
+            // The coins to merge should be used as transaction object inputs, as
+            // `TransactionData::new_programmable` used in `InternalOperation::try_into_data`,
+            // uses all coins passed as gas payment.
+            let mut iter = all_coins.iter().map(|c| c.object_ref());
+            gas_coins = iter.by_ref().take(MAX_GAS_COINS).collect();
```

Comment:
do we need to do a sort first to make sure most of the coin values are use as gas? to reduce the chances that `take(MAX_GAS_COIN)` taking all the dust results in not enough gas.

Reply:
`coin_read_api` actually returns coins ordered by balance descending: [source code snippet](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/jsonrpc_index.rs#L55C1-L62C2)

Reply:
[82c6950](https://github.com/MystenLabs/sui/pull/20283/commits/82c695002cf31c5164d70c838717fc4c7f9d27ef)